sw/device/tests/sim_dv/sysrst_ctrl_outputs_test.c

```diff
@@ -0,0 +1,167 @@
+
+enum {
+  kTestPhaseSetup = 0,
+  kTestPhaseLoopback = 1,
+  kTestPhaseOverrideSetup = 2,
+  kTestPhaseOverrideZeros = 3,
+  kTestPhaseOverrideOnes = 4,
+  kTestPhaseOverrideRelease = 5,
+  kTestPhaseOverrideAndLoopback = 6,
+  kTestPhaseDone = 7,
+};
+
+enum {
+  kAllZero = 0x0,
+  kAllOne = 0xff,
+  kLoopbackPartial = 0x5,
+  kLoopbackNumPads = 0x4,
+  kOutputNumPads = 0x8,
+  kOutputNunMioPads = 0x6,
+};
+
+static const dif_pinmux_index_t kPeripheralInputs[] = {
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonKey0In,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonKey1In,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonKey2In,
+    kTopEarlgreyPinmuxPeripheralInSysrstCtrlAonPwrbIn,
+};
+
+static const dif_pinmux_index_t kInputPads[] = {
+    kTopEarlgreyPinmuxInselIob3,
+    kTopEarlgreyPinmuxInselIob6,
+    kTopEarlgreyPinmuxInselIob8,
+    kTopEarlgreyPinmuxInselIor13,
+};
+
+static const dif_pinmux_index_t kPeripheralOutputs[] = {
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonKey0Out,
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonKey1Out,
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonKey2Out,
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonPwrbOut,
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonBatDisable,
+    kTopEarlgreyPinmuxOutselSysrstCtrlAonZ3Wakeup,
+};
+
+static const dif_pinmux_index_t kOutputPads[] = {
+    kTopEarlgreyPinmuxMioOutIob9, kTopEarlgreyPinmuxMioOutIor5,
+    kTopEarlgreyPinmuxMioOutIor6, kTopEarlgreyPinmuxMioOutIoc7,
+    kTopEarlgreyPinmuxMioOutIoc9,
+    kTopEarlgreyPinmuxMioOutIob3,  // Reassigned to an output.
+};
+
+static const dif_sysrst_ctrl_pin_t kSysrstCtrlOutputs[] = {
+    kDifSysrstCtrlPinKey0Out,           kDifSysrstCtrlPinKey1Out,
+    kDifSysrstCtrlPinKey2Out,           kDifSysrstCtrlPinPowerButtonOut,
+    kDifSysrstCtrlPinBatteryDisableOut, kDifSysrstCtrlPinZ3WakeupOut,
+    kDifSysrstCtrlPinEcResetInOut,      kDifSysrstCtrlPinFlashWriteProtectInOut,
+};
+
+// Test phase written by testbench.
+static volatile const uint8_t kTestPhase = 0;
+
+static void loopback_setup(void) {
+  for (int i = 0; i < kLoopbackNumPads; ++i) {
+    CHECK_DIF_OK(
+        dif_pinmux_input_select(&pinmux, kPeripheralInputs[i], kInputPads[i]));
+    CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kOutputPads[i],
+                                          kPeripheralOutputs[i]));
+  }
+}
+
+static void wait_next_test_phase(void) {
+  uint8_t current_phase = kTestPhase;
+  // Set WFI status for testbench synchronization,
+  // no actual WFI instruction is issued.
+  test_status_set(kTestStatusInWfi);
+  test_status_set(kTestStatusInTest);
+  IBEX_SPIN_FOR(current_phase != kTestPhase, kTestPhaseTimeoutUsec);
+}
+
+static void override_setup(uint8_t pins_to_override) {
+  for (int i = 0; i < kOutputNumPads; ++i) {
+    if ((pins_to_override >> i) & 0x1) {
+      if (i < kOutputNunMioPads) {
+        CHECK_DIF_OK(dif_pinmux_output_select(&pinmux, kOutputPads[i],
+                                              kPeripheralOutputs[i]));
+      }
```

Comment:
can we have the IO pinmux configuration in one place? We need to clean it up later as we will configure the pinmux based on use cases.
If possible, don't reuse a IO for different things

Reply:
No problem, changed.